src/utils.c

```diff
@@ -384,7 +447,17 @@ ts_date_bucket(PG_FUNCTION_ARGS)
 
 	/* convert to timestamp (NOT tz), bucket, convert back to date */
 	converted_ts = DirectFunctionCall1(date_timestamp, PG_GETARG_DATUM(1));
-	bucketed = DirectFunctionCall2(ts_timestamp_bucket, PG_GETARG_DATUM(0), converted_ts);
+	if (PG_NARGS() > 2)
+	{
+		Datum		converted_ts_origin;
+
+		converted_ts_origin = DirectFunctionCall1(date_timestamp, PG_GETARG_DATUM(2));
```

Comment:
This seems to be alot of function calls for something thats basically just integer math. Did you look at time_bucket.c? Can't you use the macro defined there for bucketing? Also I think the non-int time_bucket stuff should be moved there, there is pivotal story for that. Also I would not do the decision about offsetting at runtime. I would define different c functions for the different input configurations and use a macro for the calculation. This function is potentially executed a lot so i think its worth optimizing for that.

Reply:
Mostly Fixed. Not sure what you meant by `I would not do the decision about offsetting at runtime`